doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -0,0 +1,965 @@
+close to e.g. the one of OCaml. Types follow the simply-typed syntax of OCaml.
+
+.. productionlist:: coq
+   type             : ( `type`, ... , `type` ) `typeconst`
+                    : ( `type` * ... * `type` )
+                    : `type` -> `type`
+                    : `typevar`
+   typeconst        : ( `modpath` . )* `lident`
+   typevar          : '`lident`
+   typeparams       : ( `typevar`, ... , `typevar` )
+
+The set of base types can be extended thanks to the usual ML type
+declarations such as algebraic datatypes and records.
+
+Built-in types include:
+
+- :n:`int`, machine integers (size not specified, in practice inherited from OCaml)
+- :n:`string`, mutable strings
+- :n:`'a array`, mutable arrays
+- :n:`exn`, exceptions
+- :n:`constr`, kernel-side terms
+- :n:`pattern`, term patterns
+- :n:`ident`, well-formed identifiers
+
+Type declarations
+~~~~~~~~~~~~~~~~~
+
+One can define new types by the following commands.
+
+.. cmd:: Ltac2 Type @typeparams @lident
+   :name: Ltac2 Type
+
+   This command defines an abstract type. It has no use for the end user and
+   is dedicated to types representing data coming from the OCaml world.
+
+.. cmdv:: Ltac2 Type {? rec} @typeparams @lident := @typedef
+
+   This command defines a type with a manifest. There are four possible
+   kinds of such definitions: alias, variant, record and open variant types.
+
+   .. productionlist:: coq
+      typedef          : `type`
+                       : [ `constructordef` | ... | `constructordef` ]
+                       : { `fielddef` ; ... ; `fielddef` }
+                       : [ .. ]
+      constructordef   : `ident` [ ( `type` , ... , `type` ) ]
+      fielddef         : [ mutable ] `ident` : `type`
+
+   Aliases are just a name for a given type expression and are transparently
+   unfoldable to it. They cannot be recursive.
+
+   Variants are sum types defined by constructors and eliminated by
+   pattern-matching. They can be recursive, but the `RECFLAG` must be explicitly
+   set. Pattern-maching must be exhaustive.
+
+   Records are product types with named fields and eliminated by projection.
+   Likewise they can be recursive if the `RECFLAG` is set.
+
+   .. cmdv:: Ltac2 Type @typeparams @qualid := [ @constructordef ]
+
+      Open variants are a special kind of variant types whose constructors are not
+      statically defined, but can instead be extended dynamically. A typical example
+      is the standard `exn` type. Pattern-matching must always include a catch-all
+      clause. They can be extended by this command.
+
+Term Syntax
+~~~~~~~~~~~
+
+The syntax of the functional fragment is very close to the one of Ltac1, except
+that it adds a true pattern-matching feature, as well as a few standard
+constructions from ML.
+
+.. productionlist:: coq
+   var              : lident
+   qualid           : ( modpath . )* lident
+   constructor      : uident
+   term             : qualid
```

Comment:
There are some missing backquotes here and there is the same problem about namespacing.